File: leetcode/2642/solution.py
"""
https://leetcode.com/problems/design-graph-with-shortest-path-calculator/
2642. Design Graph With Shortest Path Calculator
"""
from typing import List


class Graph:
    def __init__(self, n: int, edges: List[List[int]]):
        self.graph = defaultdict(list)
        self.max = 1_000_001 * n
        for frm, to, cost in edges:
            self.graph[frm].append((cost, to))

    def addEdge(self, edge: List[int]) -> None:
        frm, to, cost = edge
        self.graph[frm].append((cost, to))

    def shortestPath(self, node1: int, node2: int) -> int:
        costs = defaultdict(lambda: self.max)
        seen = set()
        heap = [(0, node1)]

        while heap:
            cost, i = heappop(heap)
            if i == node2:
                # found the answer
                return cost
            # update costs[i] (min cost from node1 to i)
            costs[i] = min(costs[i], cost)
            # if we haven't visited i, then enqueue subsequent edges
            if i in seen:
                continue
            seen.add(i)
            for additional_cost, j in self.graph[i]:
                if j not in seen:
                    heappush(heap, (additional_cost + cost, j))

        return -1

File: leetcode/2642/solution.rs
/**
 * https://leetcode.com/problems/design-graph-with-shortest-path-calculator/
 * 2642. Design Graph With Shortest Path Calculator
 */
use std::collections::{BinaryHeap, HashMap, HashSet};

struct Graph {
    graph: HashMap<usize, Vec<(i32, usize)>>,
}

impl Graph {
    fn new(n: i32, edges: Vec<Vec<i32>>) -> Self {
        let graph = HashMap::new();
        let mut graph: Graph = Graph { graph };
        edges.into_iter().for_each(|edge| graph.add_edge(edge));
        graph
    }

    fn add_edge(&mut self, edge: Vec<i32>) {
        // since Rust's binary heap is max-heap we will store cost * -1
        let (from, to, cost) = (edge[0] as usize, edge[1] as usize, -edge[2]);
        self.graph
            .entry(from)
            .and_modify(|v| v.push((cost, to)))
            .or_insert(vec![(cost, to)]);
    }

    fn shortest_path(&self, node1: i32, node2: i32) -> i32 {
        let node1 = node1 as usize;
        let node2 = node2 as usize;
        let mut costs: HashMap<usize, i32> = HashMap::new();
        let mut seen: HashSet<usize> = HashSet::new();
        let mut heap: BinaryHeap<(i32, usize)> = BinaryHeap::new();
        heap.push((0, node1));

        while let Some((cost, node)) = heap.pop() {
            if node == node2 {
                // found the shortest path
                return -cost;
            }

            // update costs
            costs
                .entry(node)
                .and_modify(|v| *v = cost.min(*v))
                .or_insert(cost);

            // push neighbors
            if seen.contains(&node) {
                continue;
            }
            seen.insert(node);
            if let Some(neighbors) = self.graph.get(&node) {
                for (n_cost, neighbor) in neighbors {
                    let cost = cost + *n_cost;
                    heap.push((cost, *neighbor));
                }
            }
        }

        -1
    }
}
